Terraform apply issue with outputs

│ Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
  │ 
  │ Outputs:
  │ 
  │ vault_init_root_token = "hvs.WEwYvxVkzGJUUVbUWPczoeld"
  │ vault_init_unseal_keys = [
  │   "5c929353c12c8f2db067edb3a58c507eee85949fdc2ab70a1bb2c38bcf265523",
  │ ]

DEBU Forcefully remove container=357add9fe902e5fb0396b6012d83508397b8625e261844d3385a3533d683bdd9
INFO Please wait, still creating resources [Elapsed Time: 30.001960]
panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 124 [running]:
github.com/jumppad-labs/jumppad/pkg/config/resources/terraform.(*TerraformProvider).generateOutput(0x28786b0bd020)
        /home/nicj/code/github.com/jumppad-labs/jumppad/pkg/config/resources/terraform/provider.go:379 +0x8ea
github.com/jumppad-labs/jumppad/pkg/config/resources/terraform.(*TerraformProvider).Create(0x28786b0bd020, {0x5dd8350?, 0x28786b3a2be0?})
        /home/nicj/code/github.com/jumppad-labs/jumppad/pkg/config/resources/terraform/provider.go:84 +0x490
github.com/jumppad-labs/jumppad/pkg/jumppad.(*EngineImpl).createCallback(0x28786b3a2730, {0x5dec888, 0x28786b73e9c0})
        /home/nicj/code/github.com/jumppad-labs/jumppad/pkg/jumppad/engine.go:529 +0x3b8
github.com/jumppad-labs/hclconfig.(*Parser).process.func2({0x5dec888, 0x28786b73e9c0})
        /home/nicj/go/pkg/mod/github.com/jumppad-labs/hclconfig@v0.30.2/parser.go:1580 +0x82
github.com/jumppad-labs/hclconfig.(*Parser).process.createCallback.func4({0x5503c00?, 0x28786b73e9c0})
        /home/nicj/go/pkg/mod/github.com/jumppad-labs/hclconfig@v0.30.2/dag.go:366 +0xdfa
github.com/silas/dag.(*Walker).walkVertex(0x28786ae014a0, {0x5503c00, 0x28786b73e9c0}, 0x28786ac01a80)
        /home/nicj/go/pkg/mod/github.com/silas/dag@v0.0.0-20220518035006-a7e85ada93c5/walk.go:377 +0x2a5
created by github.com/silas/dag.(*Walker).Update in goroutine 1
        /home/nicj/go/pkg/mod/github.com/silas/dag@v0.0.0-20220518035006-a7e85ada93c5/walk.go:300 +0x1048

--- a/examples/terraform/main.hcl
+++ b/examples/terraform/main.hcl
@@ -61,6 +61,10 @@ output "third_y" {
   value = resource.terraform.configure_vault.output.third.y
 }
 
+output "list" {
+  value = resource.terraform.configure_vault.output.list
+}
+
 output "vault_secret" {
   value = resource.terraform.configure_vault.output.vault_secret
 }

--- a/examples/terraform/workspace/main.tf
+++ b/examples/terraform/workspace/main.tf
@@ -19,3 +19,7 @@ output "third" {
 output "test" {
   value = "testing123"
 }
+
+output "list" {
+  value = [var.first, var.second, var.third]
+}

--- a/pkg/config/resources/terraform/provider.go
+++ b/pkg/config/resources/terraform/provider.go
@@ -75,7 +75,10 @@ func (p *TerraformProvider) Create(ctx context.Context) error {
 	// always remove the container
 	defer p.client.RemoveContainer(id, true)
 
+	logPath := filepath.Join(utils.LogsDir(), fmt.Sprintf("terraform_%s.log", p.config.Meta.Name))
+
 	err = p.terraformApply(id)
+	p.log.Info("Terraform apply log written", "ref", p.config.Meta.ID, "log", logPath)
 	if err != nil {
 		return fmt.Errorf("unable to run apply for terraform.%s: %w", p.config.Meta.Name, err)
 	}
@@ -335,8 +338,10 @@ func (p *TerraformProvider) terraformApply(containerid string) error {
 
 	_, err := p.client.ExecuteScript(containerid, script, envs, wd, "root", "", 300, planOutput)
 
-	// write the plan output to the log
+	// write the plan output to a log file and the config
 	p.config.ApplyOutput = planOutput.String()
+	logPath := filepath.Join(utils.LogsDir(), fmt.Sprintf("terraform_%s.log", p.config.Meta.Name))
+	os.WriteFile(logPath, planOutput.Bytes(), 0644)
 	p.log.Debug("terraform apply output", "id", p.config.Meta.ID, "output", planOutput)
 
 	if err != nil {

--- a/pkg/config/resources/terraform/provider.go
+++ b/pkg/config/resources/terraform/provider.go
@@ -325,7 +325,7 @@ func (p *TerraformProvider) terraformApply(containerid string) error {
 	// add the tf vars flag if we have a file
 	script = script + tfvarFlag
 
-	script = script + `terraform output \
+	script = script + "\n" + `terraform output \
     -no-color \
     -state=/var/lib/terraform/terraform.tfstate \
     -json > /var/lib/terraform/output.json`

--- a/pkg/config/resources/terraform/provider.go
+++ b/pkg/config/resources/terraform/provider.go
@@ -8,7 +8,6 @@ import (
 	"os"
 	"path"
 	"path/filepath"
-	"reflect"
 	"strings"
 
 	"github.com/hashicorp/hcl/v2/hclwrite"
@@ -366,36 +365,58 @@ func (p *TerraformProvider) generateOutput() error {
 
 	values := map[string]cty.Value{}
 	for k, v := range output {
-		m, ok := v.(map[string]interface{})
+		m, ok := v.(map[string]any)
 		if !ok {
-			return fmt.Errorf("terraform output is not in the correct format, expected map[string]interface{} for value but got %T", v)
+			return fmt.Errorf("terraform output is not in the correct format, expected map[string]any for value but got %T", v)
 		}
 
-		value, err := convert.GoToCtyValue(m["value"])
+		ctyVal, err := goToCty(m["value"])
 		if err != nil {
-			if reflect.TypeOf(m["type"]).Kind() == reflect.Slice {
-				obj := map[string]cty.Value{}
-
-				for l, w := range m["value"].(map[string]interface{}) {
-					subvalue, err := convert.GoToCtyValue(w)
-					if err != nil {
-						p.log.Error("could not convert variable", "key", l, "value", w, "error", err)
-						return err
-					}
-
-					obj[l] = subvalue
-				}
-
-				values[k] = cty.ObjectVal(obj)
-			}
-		} else {
-			values[k] = value
+			return fmt.Errorf("terraform output %q: %w", k, err)
 		}
+
+		values[k] = ctyVal
 	}
 
 	p.config.Output = cty.ObjectVal(values)
 
 	return nil
+}
+
+// goToCty recursively converts a Go value (from JSON unmarshal) into a cty.Value.
+// It handles primitives via convert.GoToCtyValue and recurses into maps and slices.
+func goToCty(val any) (cty.Value, error) {
+	switch v := val.(type) {
+	case map[string]any:
+		obj := map[string]cty.Value{}
+		for k, w := range v {
+			sub, err := goToCty(w)
+			if err != nil {
+				return cty.NilVal, fmt.Errorf("key %q: %w", k, err)
+			}
+			obj[k] = sub
+		}
+		return cty.ObjectVal(obj), nil
+	case []any:
+		if len(v) == 0 {
+			return cty.EmptyTupleVal, nil
+		}
+		elems := make([]cty.Value, len(v))
+		for i, w := range v {
+			sub, err := goToCty(w)
+			if err != nil {
+				return cty.NilVal, fmt.Errorf("index %d: %w", i, err)
+			}
+			elems[i] = sub
+		}
+		return cty.TupleVal(elems), nil
+	default:
+		ctyVal, err := convert.GoToCtyValue(val)
+		if err != nil {
+			return cty.NilVal, fmt.Errorf("cannot convert %T to cty: %w", val, err)
+		}
+		return ctyVal, nil
+	}
 }
 
 func (p *TerraformProvider) terraformDestroy(containerid string) error {

